src/emit/substrate/target.rs

```diff
@@ -1092,19 +1092,14 @@ impl<'a> TargetRuntime<'a> for SubstrateTarget {
             binary.build_alloca(function, binary.context.i8_type().array_type(36), "salt");
         let salt_len = i32_const!(32);
 
-        if let Some(salt) = contract_args.salt {
-            binary.builder.build_store(salt_buf, salt);
-        } else {
-            let (ptr, len) = self.contract_unique_salt(binary, contract_no, ns);
-
-            binary.builder.build_store(scratch_len, i32_const!(36));
-
-            call!(
-                "seal_random",
-                &[ptr.into(), len.into(), salt_buf.into(), scratch_len.into()],
-                "random"
-            );
-        }
+        let salt = contract_args.salt.unwrap_or_else(|| {
+            call!("instantiation_nonce", &[], "instantiation_nonce")
+                .try_as_basic_value()
+                .left()
+                .unwrap()
+                .into_int_value()
+        });
+        binary.builder.build_store(salt_buf, salt);
```

Comment:
The salt_len is defined as 32 above, but we are saving a i64 into it if the salt is not set.

Reply:
The type of salt when specified in the contract is `uint256`, see `fn parse_call_args()`. It might be nicer to make this consistent (although it should remain `uint256` for the evm target, still hope to get that working one day).